src/Container/TarballReadFs.hs

```diff
@@ -0,0 +1,297 @@
+{-# LANGUAGE GADTs #-}
+
+module Container.TarballReadFs (
+  runTarballReadFSIO,
+) where
+
+import Codec.Archive.Tar (
+  Entry (entryContent),
+  EntryContent (
+    BlockDevice,
+    CharacterDevice,
+    Directory,
+    HardLink,
+    NamedPipe,
+    NormalFile,
+    OtherEntryType,
+    SymbolicLink
+  ),
+ )
+import Codec.Archive.Tar.Entry (entryTarPath, fromLinkTargetToPosixPath)
+import Codec.Archive.Tar.Index (TarEntryOffset, hReadEntry)
+import Container.Tarball (filePathOf)
+import Control.Carrier.Simple (interpret)
+import Control.Effect.Exception (throw)
+import Control.Effect.Lift (Has, Lift)
+import Control.Exception (
+  Exception (fromException, toException),
+  IOException,
+  SomeAsyncException (SomeAsyncException),
+  catch,
+  throwIO,
+ )
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as ByteString
+import Data.FileTree.IndexFileTree (
+  SomeFileTree,
+  doesDirExist,
+  doesFileExist,
+  lookupDir,
+  lookupFileRef,
+  resolveSymLinkRef,
+ )
+import Data.Hashable (hash)
+import Data.Set qualified as Set
+import Data.String.Conversion (
+  ConvertUtf8 (decodeUtf8),
+  LazyStrict (toStrict),
+  ToText (toText),
+  toString,
+ )
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Effect.ReadFS (
+  DirID (..),
+  ReadFSErr (CurrentDirError, FileReadError, ListDirError, ResolveError),
+  ReadFSF (
+    DoesDirExist,
+    DoesFileExist,
+    GetCurrentDir,
+    GetIdentifier,
+    ListDir,
+    ReadContentsBS',
+    ReadContentsBSLimit',
+    ReadContentsText',
+    ResolveDir',
+    ResolveFile',
+    ResolvePath
+  ),
+  ReadFSIOC,
+  SomePath (SomeDir, SomeFile),
+  catchingIO,
+ )
+import Path (Abs, Dir, File, SomeBase (..))
+import Path.Internal (Path (Path))
+import System.IO (IOMode (ReadMode), withFile)
+
+maxHopsOf20 :: Int
+maxHopsOf20 = 20
```

Comment:
I don't really feel that strongly because this is used in one place and limited to one file. In general I wouldn't include the number in the name because it defeats the purpose of the indirection IMO.